core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -352,6 +355,7 @@ public TableMetadata updateSchema(Schema newSchema, int newLastColumnId) {
         currentSnapshotId, snapshots, snapshotLog, addPreviousFile(file, lastUpdatedMillis));
   }
 
+  // Input newPartitionSpec's partition field IDs should have already been recomputed
```

Comment:
#903 added `formatVersion` to metadata, so now we can add a validation here when the table is v1 that the spec IDs are assigned as expected:

```
ValidationException.check(formatVersion > 1 || hasSequentialIds(spec),
    "Spec does not use sequential IDs that are required in v1: %s", spec);
```